src/reports_updater.rs

```diff
@@ -96,45 +104,242 @@ impl TcnMatcherRayon {
+        let merged = self.merge_with_db(tcns.clone())?;
+
+        self.tcn_dao.overwrite(merged)?;
+
+        tcns.clear();
+
+        Ok(())
+    }
+
+    pub fn push(&self, tcn: ObservedTcn) {
+        let res = self.tcns_batch.lock();
+        let mut tcns = expect_log!(res, "Couldn't lock tcns batch");
+        
+        // TCNs in batch are merged to save memory and simplify processing / reduce logs.
+        let merged_tcn = match tcns.get(&tcn.tcn.0) {
+            Some(existing_tcn) => match Self::merge_tcns(&self.exposure_grouper, existing_tcn.to_owned(), tcn.clone()) {
+                Some(merged) => merged,
+                None => tcn
+            },
+            None => tcn
+        };
+        tcns.insert(merged_tcn.tcn.0, merged_tcn);
+    }
+
+    // Used only in tests
+    #[allow(dead_code)]
+    pub fn len(&self) -> Result<usize, ServicesError> {
+        match self.tcns_batch.lock() {
+            Ok(tcns) => Ok(tcns.len()),
+            Err(_) => Err(ServicesError::General("Couldn't lock tcns batch".to_owned()))
+        }
+    }
+
+    // Retrieves possible existing exposures from DB with same TCNs and does an in-memory merge.
+    fn merge_with_db(&self, tcns: HashMap<[u8; 16], ObservedTcn>) -> Result<Vec<ObservedTcn>, ServicesError> {
+        let tcns_vec: Vec<ObservedTcn> = tcns.iter().map(|(_, observed_tcn)| observed_tcn.clone()).collect();
+
+        let mut db_tcns = self
+            .tcn_dao
+            .find_tcns(tcns_vec.clone().into_iter().map(|tcn| tcn.tcn).collect())?;
+        db_tcns.sort_by_key(|tcn| tcn.contact_start.value);
+        
+        let db_tcns_map: HashMap<[u8; 16], Vec<ObservedTcn>> = Self::to_hash_map(db_tcns);
+
+        Ok(tcns_vec.into_iter().map(|tcn|
+            // Values in db_tcns_map can't be empty: we built the map based on existing TCNs
+            Self::determine_tcns_to_write(self.exposure_grouper.clone(), db_tcns_map.clone(), tcn)
+        ).flatten().collect())
+    }
+
+    // Expects:
+    // - Values in db_tcns_map not empty
+    // - db_tcns_map sorted by contact_start (ascending)
+    fn determine_tcns_to_write(exposure_grouper: ExposureGrouper, db_tcns_map: HashMap<[u8; 16], Vec<ObservedTcn>>, tcn: ObservedTcn) -> Vec<ObservedTcn> {
+        let db_tcns = db_tcns_map.get(&tcn.tcn.0);
+
+        match db_tcns {
+            // Matching exposures in DB
+            Some(db_tcns) => {
+                if let Some(last) = db_tcns.last() {
+
+                    // If contiguous to last DB exposure, merge with it, otherwise append.
+                    let tail = match Self::merge_tcns(&exposure_grouper, last.to_owned(), tcn.clone()) {
+                        Some(merged ) => vec![merged],
+                        None => vec![last.to_owned(), tcn]
+                    };
+                    let mut head: Vec<ObservedTcn> = db_tcns.to_owned().into_iter().take(db_tcns.len() - 1).collect();
```

Comment:
We could avoid iterating over the vector and just replace the last element: https://doc.rust-lang.org/std/mem/fn.replace.html

Reply:
Unclear how `mem::replace` helps. It seems to replace the complete vector. `db_tcns` is also unlikely to contain more than a couple of elements (these are the stored, averaged exposures for an individual TCN) so the performance is negligible.

Reply:
mem::replace works on a vector element. 
Good point regarding small vector size.